pkg/stanza/decoder/decoder.go

```diff
@@ -0,0 +1,81 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package decoder // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/decoder"
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/ianaindex"
+	"golang.org/x/text/encoding/unicode"
+	"golang.org/x/text/transform"
+)
+
+type Decoder struct {
+	encoding     encoding.Encoding
+	decoder      *encoding.Decoder
+	decodeBuffer []byte
+}
+
+// New wraps a character set encoding and creates a reusable buffer to reduce allocation.
+// Decoder is not thread-safe and must not be used in multiple goroutines.
+func New(enc encoding.Encoding) *Decoder {
+	return &Decoder{
+		encoding:     enc,
+		decoder:      enc.NewDecoder(),
+		decodeBuffer: make([]byte, 1<<12),
+	}
+}
+
+// Decode converts the bytes in msgBuf to UTF-8 from the configured encoding.
+func (d *Decoder) Decode(msgBuf []byte) ([]byte, error) {
+	for {
+		d.decoder.Reset()
+		nDst, _, err := d.decoder.Transform(d.decodeBuffer, msgBuf, true)
+		if err == nil {
+			return d.decodeBuffer[:nDst], nil
+		}
+		if errors.Is(err, transform.ErrShortDst) {
+			d.decodeBuffer = make([]byte, len(d.decodeBuffer)*2)
+			continue
+		}
+		return nil, fmt.Errorf("transform encoding: %w", err)
+	}
+}
+
+var encodingOverrides = map[string]encoding.Encoding{
```

Comment:
I realize we're just copying this over, but since we're creating a new package for this would it make since to have the keys alphabetical order, or sorted in some way? Feel free to ignore if you don't think it'd help.

Reply:
Good point. I've sorted them roughly in what seems like a more logical order.